CRD supports multi-version Schema, Subresources and AdditionalPrintColumns

roycaihw opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Part of the CRD conversion webhook design: [customresource-conversion-webhook.md#top-level-fields-to-per-version-fields](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/customresource-conversion-webhook.md#top-level-fields-to-per-version-fields) 
Main CRD conversion webhook PR: https://github.com/kubernetes/kubernetes/pull/67006
This PR is a combination of two previous separate PRs (https://github.com/kubernetes/kubernetes/pull/67553, https://github.com/kubernetes/kubernetes/pull/67521) following the updated KEP design: https://github.com/kubernetes/community/pull/2737

**Special notes for your reviewer**:
1. Please review the integration test change (in a standalone commit) with whitespace change ignored.
2. This PR currently does ~~redundant~~ **pre-creation/update-dropping-field** ~~and **failing-validation**~~ for disabled feature-gated fields, please refer to the related topic in [sig-api-machinery meeting (Oct 24)](https://docs.google.com/document/d/1x9RNaaysyO0gXHIr1y50QFbiL1x8OWnk2v3XnrdkT5Y/edit#heading=h.vfnf5phmk3ad) for more information.
3. ~~Two subresources integration tests (with **TODO** comments) will be extended to test per-version subresources after https://github.com/kubernetes/kubernetes/pull/70087 fixes the unstructured conversion bug.~~ Updated

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE".
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
2. 
-->
```release-note
CRD supports multi-version Schema, Subresources and AdditionalPrintColumns (NOTE that CRDs created prior to 1.13 populated the top-level additionalPrinterColumns field by default. To apply an update that changes to per-version additionalPrinterColumns, the top-level additionalPrinterColumns field must be explicitly set to null).
```

/sig api-machinery
/area custom-resources
/cc @sttts @mbohlool @liggitt

roycaihw commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -182,7 +260,7 @@ func ValidateCustomResourceDefinitionSpec(spec *apiextensions.CustomResourceDefi
 
 	for i := range spec.AdditionalPrinterColumns {
-		if errs := ValidateCustomResourceColumnDefinition(&spec.AdditionalPrinterColumns[i], fldPath.Child("columns").Index(i)); len(errs) > 0 {
+		if errs := ValidateCustomResourceColumnDefinition(&spec.AdditionalPrinterColumns[i], fldPath.Child("additionalPrinterColumns").Index(i)); len(errs) > 0 {
```
IIUC the validation error shows the jsonpath of the problematic field, so this should've been `additionalPrinterColumns` instead of `columns`

roycaihw commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -94,10 +110,36 @@ func (strategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+	// fields. If the old CRD already uses per-version fields, the CRD is allowed to continue
+	// use per-version fields.
+	if !utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceWebhookConversion) &&
+		!hasPerVersionField(oldCRD.Spec.Versions) {
```
I'm okay with auto-clearing when gate is off to make the apiserver close to if the feature doesn't exist (auto-clearing on CREATE has been approved in the KEP, but for UPDATE it wasn't clear to me). I still kept the redundant validation for now.

cc @mbohlool @liggitt

roycaihw commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go:
```diff
@@ -17,9 +17,20 @@ limitations under the License.
 )
 
+// CustomResourceColumns masks the slice so protobuf can preserve empty slice
+// +protobuf.nullable=true
```
@liggitt Do we need to test the round-tripping?

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -20,6 +20,9 @@ import (
 )
 
+// CustomResourceColumns masks the slice so protobuf can preserve empty slice
+type CustomResourceColumns []CustomResourceColumnDefinition
```
remind me why we care about distinguishing nil and [] here?

roycaihw replied:
The concern was brought up in https://github.com/kubernetes/kubernetes/pull/67521#discussion_r215220972. My understanding is that we need to make protobuf preserve [], to allow overriding with an empty list. I didn't change the top-level field because it would be a breaking change.

liggitt replied:
rereading that thread, it looks like the feedback there was to avoid making the distinction between nil and [] (to not do this nullable typedef). I think the `len(...) == 0` checks for emptiness are sufficient. Were there specific places where we needed to be able to tell nil from [] for some reason?

roycaihw replied:
Thanks for the explanation :) No the apiserver doesn't need to distinguish nil and [] (and defaulting happens for both cases).

liggitt replied:
great, then let's drop this and just go with `[]CustomResourceColumnDefinition`

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -61,6 +73,21 @@ type CustomResourceDefinitionVersion struct {
+	// Schema describes the schema for CustomResource used in validation, pruning, and defaulting.
+	// Top-level and per-version schemas are mutually exclusive.
+	// Per-version schemas may not all be set to identical values (top-level validation schema should be used instead)
+	// +optional
```
for all the alpha, feature-gated fields, add both of those details to the documentation. For example:

```
// This field is alpha-level and is only honored by servers that enable the XYZ feature.
```

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -207,6 +285,91 @@ func ValidateCustomResourceDefinitionSpecUpdate(spec, oldSpec *apiextensions.Cus
+	}
+	value := versions[0].AdditionalPrinterColumns
+	for _, v := range versions {
+		if v.AdditionalPrinterColumns == nil || !reflect.DeepEqual(v.AdditionalPrinterColumns, value) {
```
should we check len(...)==0 instead of nil? should we use semantic equality rather than reflect.DeepEqual?

roycaihw replied:
By semantic equality you mean for two column slices, we should consider them to be the same if they have same elements but in different order?

liggitt replied:
I mean to use `apiequality.Semantic.DeepEqual()`. It equates nil and empty lists, and does a few other semantic comparisons (like the absolute value of resource.Quantity fields, rather than their literal value, so 1M == 1000)

nikopen commented:
/milestone v1.13

roycaihw commented:
@liggitt Addressed comments and updated integration test (I had to rebase to pick up the conversion bug fix, but the first five commits didn't change). PTAL :)

roycaihw commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go:
```diff
@@ -66,6 +72,24 @@ type CustomResourceDefinitionVersion struct {
+	// Per-version schemas may not all be set to identical values (top-level validation schema should be used instead)
+	// This field is alpha-level and is only honored by servers that enable the CustomResourceWebhookConversion feature.
+	// +optional
+	Schema *JSONSchemaProps `json:"schema,omitempty" protobuf:"bytes,4,opt,name=schema"`
```
My understanding is that for the top-level field, we were planning on having [a list of validation methods](https://github.com/kubernetes/kubernetes/blob/76234a31b0491134b6c08f65b7eeda376dcaf61f/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go#L220-L221) for CRs, so we defined a struct and made `openAPIV3Schema` nested. 

Now it becomes clear that we are not adding new methods to the `CustomResourceValidation`, so we simplify the per-version schema to just `JSONSchemaProps`. 

also cc @mbohlool @lavalamp on the related conversation: https://github.com/kubernetes/kubernetes/issues/67840

liggitt replied:
> Now it becomes clear that we are not adding new methods to the `CustomResourceValidation`, so we simplify the per-version schema to just `JSONSchemaProps`.
> 
> also cc @mbohlool @lavalamp on the related conversation: #67840

#67840 actually proposes supporting a schema format other than openapi v3, which is exactly the use case for nesting openapi v3 as one of potentially multiple schema types. I think we should make the per-version schema structs in v1beta1 match the top-level structs, so that we can round-trip to whatever we end up with in v1, and have a path to transition to a different schema if that's where #67840 ends up. @lavalamp @sttts do you agree?

roycaihw replied:
Yes. In the KEP we proposed dropping the nested struct and going with just openapi v3, then in #67840 we brought up the idea around replacing openapi v3. I'm happy to go with the nested struct if we think that will make it easier to support another schema format, before we merge this PR.

mbohlool replied:
I am fine with keeping top level struct format with nested openapi v3. While last proposal on #67840 is to keep a more restricted version of openapi v3 but keeping the options open make more sense.

roycaihw replied:
More related context: https://github.com/kubernetes/kubernetes/issues/64483

I'm going to add a commit with
```
Schema *CustomResourceValidation `json:"schema,omitempty" protobuf:"bytes,4,opt,name=schema"`
```
first, which may look like:
```
spec:
  versions:
    - name: v1
      schema:
        openAPIV3Schema: ...       // covers schema, defaults and static validation
        anotherSchemaLanguage: ... // to be added in future?
      prune: true
```

roycaihw commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -443,19 +448,26 @@ func (r *crdHandler) getOrCreateServingInfoFor(crd *apiextensions.CustomResource
-		validator, _, err := apiservervalidation.NewSchemaValidator(crd.Spec.Validation)
+		validationSchema, err := getCRDSchemaForVersion(&crd.Spec, v.Name)
+		if err != nil {
+			return nil, err
```
Good point. Yes we shouldn't return the full CRD spec to a CR client when `ServeHTTP`. I think the same comment also applies to https://github.com/kubernetes/kubernetes/pull/70211/files/956879420560759244d2210acff668902ef20780#diff-2f720a7e61a4ddfcfe9a9e353be8784bR225.

May I ask what kind of generic error message do you have in mind? While the error can be triggered by spec with malformed subresources/schema/columns, the incoming HTTP request isn't necessarily asking for subresources-- it's called here in `ServeHTTP`: https://github.com/kubernetes/kubernetes/blob/a49e4e128e8d5ca8405fb2a26523294d65bf42e8/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L207

So something like `the server could not properly server the CRD subresources/schema/columns` (corresponding to which field might be malformed)?

liggitt replied:
yeah, something like that would be fine

lavalamp commented:
I'm a big fan of leaving open the door for another schema format.

On Tue, Oct 30, 2018 at 4:18 PM Haowei Cai (Roy) <notifications@github.com>
wrote:

> *@roycaihw* commented on this pull request.
> ------------------------------
>
> In
> staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go
> <https://github.com/kubernetes/kubernetes/pull/70211#discussion_r229519673>
> :
>
> > @@ -66,6 +72,24 @@ type CustomResourceDefinitionVersion struct {
>  	// Storage flags the version as storage version. There must be exactly one
>  	// flagged as storage version.
>  	Storage bool `json:"storage" protobuf:"varint,3,opt,name=storage"`
> +	// Schema describes the schema for CustomResource used in validation, pruning, and defaulting.
> +	// Top-level and per-version schemas are mutually exclusive.
> +	// Per-version schemas may not all be set to identical values (top-level validation schema should be used instead)
> +	// This field is alpha-level and is only honored by servers that enable the CustomResourceWebhookConversion feature.
> +	// +optional
> +	Schema *JSONSchemaProps `json:"schema,omitempty" protobuf:"bytes,4,opt,name=schema"`
>
> More related context: #64483
> <https://github.com/kubernetes/kubernetes/issues/64483>
>
> I'm going to add a commit with
>
> Schema *CustomResourceValidation `json:"schema,omitempty" protobuf:"bytes,4,opt,name=schema"`
>
> first, which may look like:
>
> spec:
>   versions:
>     - name: v1
>       schema:
>         openAPIV3Schema: ...       // covers schema, defaults and static validation
>         anotherSchemaLanguage: ... // to be added in future?
>       prune: true
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/70211#discussion_r229519673>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglq-Vl38t8cITMyUEwnN_EDM-8FNiks5uqN4sgaJpZM4X5U_r>
> .
>

mbohlool commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/helpers.go:
```diff
@@ -0,0 +1,74 @@
+		if v.Schema != nil {
+			return v.Schema, nil
+		}
+		return crd.Spec.Validation, nil
```
are we doing override here? we agreed not to do overrides. That means in getCRDSchemaForVersion we should call the other helper to see if there is any per-version Schema. If there is we should return `v.Schema` even if it is `nil`.

roycaihw replied:
No, it's not doing override. I think the auto-clear strategy makes sure that both ways work the same. In the case that you described (`v.Schema` is `nil` while there exists other per-version schema), `crd.Spec.Validation` will also be `nil`, so `getCRDSchemaForVersion` returns `nil` in the end.

mbohlool replied:
You are right. Still to anybody reading this code, it look like we are overriding. It would be a cleaner code if we apply the logic here too and it wouldn't hurt.

mbohlool replied:
or you can check if top level field is not `nil` and return it at the start of the method with a comment explaining why if it is not nil, it means all per versions will be. I still prefer the first approach though :)

mbohlool commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/helpers.go:
```diff
@@ -0,0 +1,74 @@
+		if v.Subresources != nil {
+			return v.Subresources, nil
+		}
+		return crd.Spec.Subresources, nil
```
ditto

mbohlool commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/helpers.go:
```diff
@@ -0,0 +1,74 @@
+		if len(v.AdditionalPrinterColumns) > 0 {
+			return v.AdditionalPrinterColumns, nil
+		}
+		return crd.Spec.AdditionalPrinterColumns, nil
```
ditto

mbohlool commented on staging/src/k8s.io/apiextensions-apiserver/pkg/features/kube_features.go:
```diff
@@ -40,6 +40,12 @@ const (
 	// CustomResourceSubresources defines the subresources for CustomResources
 	CustomResourceSubresources utilfeature.Feature = "CustomResourceSubresources"
+
+	// owner: @mbohlool
```
add yourself as owner here too please.

roycaihw replied:
Done

roycaihw commented:
/test pull-kubernetes-e2e-kops-aws

mbohlool commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/helpers.go:
```diff
@@ -0,0 +1,95 @@
+		if version != v.Name {
+			continue
+		}
+		if hasPerVersionSchema(crd.Spec.Versions) {
```
you can do this check earlier before for loop and return crd.Spec.Validation as early as possible then you will have less nested blocks.

liggitt replied:
agree that unnesting like this would improve readability. up to you if you want to unnest like this:

```go
if !hasPerVersionSchema(crd.Spec.Versions) {
  return crd.Spec.Validation, nil
}
if crd.Spec.Validation != nil {
  return nil, fmt.Errorf("malformed CustomResourceDefinition %s version %s: top-level and per-version schemas must be mutual exclusive", crd.Name, version)
}
for _, v := range crd.Spec.Versions {
  if version == v.Name {
    return v.Schema, nil
  }
}
return nil, fmt.Errorf("version %s not found in CustomResourceDefinition: %v", version, crd.Name)
```

mbohlool commented:
Looks good. please squash commits so we can do a final round of review per commit.

roycaihw commented:
@mbohlool @liggitt Squashed. PTAL

/test pull-kubernetes-integration

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -443,19 +444,20 @@ func (r *crdHandler) getOrCreateServingInfoFor(crd *apiextensions.CustomResource
-		if utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceSubresources) && crd.Spec.Subresources != nil && crd.Spec.Subresources.Status != nil {
-			statusSpec = crd.Spec.Subresources.Status
-
+		subresources := getCRDSubresourcesForVersionOrDie(crd, v.Name)
```
not sure why OrDie variants are being used here... why wouldn't we return an error?

mbohlool replied:
Are you against all DoXOrDie methods or only here? What is the difference?

liggitt replied:
using them for programming errors is still a little questionable when there's an easy way to propagate an error, but is more reasonable.

this is using them to fail on data errors, which I think is a bad idea.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -443,19 +444,20 @@ func (r *crdHandler) getOrCreateServingInfoFor(crd *apiextensions.CustomResource
 		creator := unstructuredCreator{}
 
-		validator, _, err := apiservervalidation.NewSchemaValidator(crd.Spec.Validation)
+		validationSchema := getCRDSchemaForVersionOrDie(crd, v.Name)
```
same question... why do we want to panic here instead of returning an error?

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/helpers.go:
```diff
@@ -0,0 +1,101 @@
+		}
+		return v.Schema
+	}
+	// Given that we call HasServedCRDVersion when serving CRD, this should not happen.
```
I think returning an error is better than panicking. Same comment applies in other methods

mbohlool replied:
We do have similar cases were we assume things. E.g if default makes sure a pointer id never nil, we don't check for nil repeatedly in the code. I think an error chain for something that cannot happened does not make sense.

liggitt replied:
defaulting is different, since it is run every time we read from etcd or from an incoming request. this is depending on data enforced one time in the past in validation.

mbohlool replied:
> this is depending on data enforced one time in the past in validation.

That is not true in this case. It is not validation we depend on, it is a check at the start of each request that make sure the request version exists. So it is at the same level as defaulting which happens every request.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/helpers.go:
```diff
@@ -20,14 +20,15 @@ import (
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/client-go/dynamic"
+
```
revert this import change

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/helpers.go:
```diff
@@ -92,3 +93,81 @@ func updateCustomResourceDefinitionWithRetry(client clientset.Interface, name st
 }
+
+// getCRDSchemaForVersionOrDie returns the validation schema for given version in given CRD.
+func getCRDSchemaForVersionOrDie(crd *apiextensionsv1beta1.CustomResourceDefinition, version string) *apiextensionsv1beta1.CustomResourceValidation {
```
prefer propagating errors to taking down the server

mbohlool replied:
The error should never happened here. Earlier in the serveHTTP function we make sure the requestVersion is valid

mbohlool replied:
Also this will not bring down the server. I am sure somewhere in http handler we catch panics and fail the request.

mbohlool replied:
My question is why it is preferable when the error is impossible to happened. There is no scenareo this error can be happened. If it happened, these are two options:
- Return a soft error which fail the request with something user won't understand (logically this cannot happened, so there is no good explanation for the end user for this)
- Return internal error with a panic logged into the api server logs which get attention from admins and ultimately us.

The chance of a panic getting our attention is more and the end result for the user is the same so I would prefer panic here. Also If we don't prefer panic here, I would say no other DoXOrDie make sense.

roycaihw commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/defaults.go:
```diff
@@ -66,7 +66,17 @@ func SetDefaults_CustomResourceDefinitionSpec(obj *CustomResourceDefinitionSpec)
+				}
+			}
+		}
+	} else if len(obj.AdditionalPrinterColumns) == 0 {
```
@mbohlool had a good point on this defaulting code:

- For existing CRDs, we (the existing code) have defaulted the top-level columns to be non-nil. 
- For new CRDs (after upgrade), the top-level columns will be defaulted (new code in this PR) if a user doesn't specify any top-level/per-version columns.

In both cases, if a user decides to patch per-version columns after upgrade (with gate enabled), the PATCH request will fail the mutual-exclusive validation-- the user has to do an UPDATE.

We basically cannot avoid the first case. For the second case, @mbohlool suggested that we can do logical defaulting to improve the user experience.

I feel it's unfair to ask users to send UPDATE requests in the second case. I'm going to add the logical defaulting code.  For the first case, we have to add documentation on that.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -174,6 +174,9 @@ type CustomResourceDefinitionVersion struct {
 	// This field is alpha-level and is only honored by servers that enable the CustomResourceWebhookConversion feature.
+	// NOTE: for CRDs created prior to 1.13 release, apiserver has defaulted the top-level additionalPrinterColumns
+	// to be non-nil. Therefore if you want to use the per-version additionalPrinterColumns, you have to use UPDATE
+	// instead of PATCH to clear the top-level additionalPrinterColumns, to avoid mutual-exclusive validation error.
```
a patch request can set the top-level columns to null explicitly at the same time it sets the per-version columns, right? I don't think this has to dictate using UPDATE

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go:
```diff
@@ -181,6 +181,9 @@ type CustomResourceDefinitionVersion struct {
 	// This field is alpha-level and is only honored by servers that enable the CustomResourceWebhookConversion feature.
+	// NOTE: for CRDs created prior to 1.13 release, apiserver has defaulted the top-level additionalPrinterColumns
+	// to be non-nil. Therefore if you want to use the per-version additionalPrinterColumns, you have to use UPDATE
+	// instead of PATCH to clear the top-level additionalPrinterColumns, to avoid mutual-exclusive validation error.
```
same comment about patch being able to clear the top-level field

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/table_test.go:
```diff
@@ -256,6 +257,48 @@ func TestTableGet(t *testing.T) {
+		t.Fatal(err)
+	}
+
+	// One should be able to patch the CRD to use per-version columns. The top-level columns
```
add a patch test that updates a CRD that only has the defaulted top-level columns by explicitly setting the top-level columns to `null` and setting per-version columns instead?

roycaihw commented:
/test pull-kubernetes-e2e-kops-aws

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -443,19 +449,28 @@ func (r *crdHandler) getOrCreateServingInfoFor(crd *apiextensions.CustomResource
+		validationSchema, err := getCRDSchemaForVersion(crd, v.Name)
+		if err != nil {
+			utilruntime.HandleError(err)
+			return nil, fmt.Errorf("the server could not properly server the CRD schema")
```
typo: could not properly *serve* (applies in several places)

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -149,6 +158,27 @@ type CustomResourceDefinitionVersion struct {
+	// Top-level and per-version columns are mutually exclusive.
+	// Per-version columns may not all be set to identical values (top-level columns should be used instead)
+	// This field is alpha-level and is only honored by servers that enable the CustomResourceWebhookConversion feature.
+	// NOTE: for CRDs created prior to 1.13 release, apiserver has defaulted the top-level additionalPrinterColumns
```
suggest: "CRDs created prior to 1.13 populated the top-level `additionalPrinterColumns` field by default. To apply an update that changes to per-version `additionalPrinterColumns`, the top-level `additionalPrinterColumns` field must be explicitly set to `null`"

liggitt reviewed (commented):
lgtm functionally, a couple nits on comments and structure. go ahead and squash in the comment commit

roycaihw commented:
@liggitt Thanks for the suggestions. I amended latest change to the last commit. Please take a look

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -149,6 +158,24 @@ type CustomResourceDefinitionVersion struct {
 	Storage bool
+	// Schema describes the schema for CustomResource used in validation, pruning, and defaulting.
+	// Top-level and per-version schemas are mutually exclusive.
+	// Per-version schemas may not all be set to identical values (top-level validation schema should be used instead)
```
What does "may not" mean here?

roycaihw replied:
Updated to "must not"

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -149,6 +158,24 @@ type CustomResourceDefinitionVersion struct {
+	// This field is alpha-level and is only honored by servers that enable the CustomResourceWebhookConversion feature.
+	// +optional
+	Schema *CustomResourceValidation
+	// Subresources describes the subresources for CustomResources
```
CustomResource<no s>

roycaihw replied:
Updated. 

This was copied from the top-level Subresources comment. I updated the top-level one as well. We have similar comments for other fields (e.g. `Validation describes the validation methods for CustomResources`). It might worth another PR to figure out the scope and do the clean up.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -250,6 +283,119 @@ func ValidateCustomResourceDefinitionSpecUpdate(spec, oldSpec *apiextensions.Cus
+		if version != v.Name {
+			continue
+		}
+		if hasPerVersionSubresources(crd.Versions) {
```
odd to have this inside the loop

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -96,10 +112,36 @@ func (strategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+	// fields. If the old CRD already uses per-version fields, the CRD is allowed to continue
+	// use per-version fields.
+	if !utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceWebhookConversion) &&
+		!hasPerVersionField(oldCRD.Spec.Versions) {
```
why the hasPerVersionField call?

roycaihw replied:
We allow a CRD to continue updating per-version fields if it already has set any one of them (schema/subresources/columns), even when the gate is off.

It's around the topic of how feature gate should work.

sttts replied:
missed oldCRD in the check and newCRD in the loop. Sgtm.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/helpers.go:
```diff
@@ -0,0 +1,104 @@
+
+// getCRDSubresourcesForVersion returns the subresources for given version in given CRD.
+func getCRDSubresourcesForVersion(crd *apiextensions.CustomResourceDefinition, version string) (*apiextensions.CustomResourceSubresources, error) {
+	if crd.Spec.Subresources != nil && hasPerVersionSubresources(crd.Spec.Versions) {
```
is this check necessary? We have validation in place.

roycaihw replied:
You're right. We have validation that should prevent this from happening and I doubt if we will ever run into this. The reason we have this here is to be extra safe. We have two "impossible errors" here:

- Top-level and per-version Xs are set at the same time: validation should prevent this; it may only happen if some admin writes to etcd directly to make the data corrupted
- The input version is not found: we have function that checks if `version` exists at early stage in cr_handler; it may only happen if we programmed it wrong

We discussed about both cases and decided to be extra safe by 
1. running `utilruntime.HandleError` that logs the error 
2. returning 500 internal server error

in both cases.

sttts replied:
Certainly does not hurt. Sgtm.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/helpers.go:
```diff
@@ -0,0 +1,104 @@
+	return nil, fmt.Errorf("version %s not found in CustomResourceDefinition: %v", version, crd.Name)
+}
+
+// getCRDSubresourcesForVersion returns the subresources for given version in given CRD.
```
drop the CRD infix

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/helpers.go:
```diff
@@ -0,0 +1,104 @@
+		if version != v.Name {
+			continue
+		}
+		if hasPerVersionSubresources(crd.Spec.Versions) {
```
move outside of loop

roycaihw replied:
Unnesting helpers.go was addressed in the 6th commit. Mark similar comments as resolved.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/helpers.go:
```diff
@@ -0,0 +1,104 @@
+		if version != v.Name {
+			continue
+		}
+		if hasPerVersionColumns(crd.Spec.Versions) {
```
move outside of loop

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -220,10 +220,16 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	subresources, err := getCRDSubresourcesForVersion(crd, requestInfo.APIVersion)
+	if err != nil {
+		utilruntime.HandleError(err)
+		http.Error(w, "the server could not properly server the CRD subresources", http.StatusInternalServerError)
```
these are not CRD subresources, but CustomResource (CR) subresources. This applies to a number of places.

roycaihw replied:
Updated.

nit: the problematic object was the CRD. Maybe we should call it "could not properly serve subresources defined in CRD"

sttts replied:
"defined by CRDs" is also fine, either way :+1:

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/helpers.go:
```diff
@@ -92,3 +92,84 @@ func updateCustomResourceDefinitionWithRetry(client clientset.Interface, name st
+
+// getCRDSchemaForVersion returns the validation schema for given version in given CRD.
+func getCRDSchemaForVersion(crd *apiextensionsv1beta1.CustomResourceDefinition, version string) (*apiextensionsv1beta1.CustomResourceValidation, error) {
+	if crd.Spec.Validation != nil && hasPerVersionSchema(crd.Spec.Versions) {
```
why those checks here everywhere?

roycaihw replied:
Please see above https://github.com/kubernetes/kubernetes/pull/70211#discussion_r231240301

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/helpers.go:
```diff
@@ -92,3 +92,84 @@ func updateCustomResourceDefinitionWithRetry(client clientset.Interface, name st
+		if version != v.Name {
+			continue
+		}
+		if hasPerVersionSchema(crd.Spec.Versions) {
```
move out of loop

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -174,6 +174,9 @@ type CustomResourceDefinitionVersion struct {
 	// Top-level and per-version columns are mutually exclusive.
 	// Per-version columns may not all be set to identical values (top-level columns should be used instead)
 	// This field is alpha-level and is only honored by servers that enable the CustomResourceWebhookConversion feature.
+	// NOTE: CRDs created prior to 1.13 populated the top-level additionalPrinterColumns field by default. To apply an
```
The "Defaults to a created-at column." is not really correct, is it? It only defaults if at least one version has additionalPrinterColumns. Otherwise, the global field is defaulted.

roycaihw replied:
> It only defaults if at least one version has additionalPrinterColumns. Otherwise, the global field is defaulted

That was the previous approach. @mbohlool [pointed out](https://github.com/kubernetes/kubernetes/pull/70211#discussion_r230518621) that in this way, we have the problem that if a user first created a CRD without specifying global/per-version columns, and then tries to patch per-version columns, the patch request will fail the mutual-exclusive validation.

We decided to drop the validation and only do logical default to improve the user experience for new CRDs. For CRDs created prior to 1.13, we cannot avoid the issue so we document in the API and in release note. The change is in 6th commit.

sttts replied:
Found it. Is it written in the release notes? Is it a breaking API change to change (remove) defaults? Strictly speaking I think it is. /cc @liggitt

liggitt replied:
> Is it a breaking API change to change (remove) defaults? Strictly speaking I think it is.

Making the existing defaulted field mutually exclusive with the per-version fields (which is required if we want to only have per-version fields in v1) unavoidably requires changing the defaulting behavior.

I can't think of any scenarios this change causes problems with:
* existing persisted CRDs retain current behavior (with respect to what columns are returned by server-side printing)
* existing CRD manifests retain current behavior when submitted to a new API server (with respect to what columns are returned by server-side printing)
* skewed HA API servers across a 1.12/1.13 boundary would serve a CRDs without this field identically
* the field is marked as optional in the openapi schema, so generated clients will tolerate it being unset

I'm comfortable proceeding as is

roycaihw commented:
/test pull-kubernetes-integration

roycaihw commented:
@sttts Comments addressed. Please take a look :)

@liggitt I interpreted "squash in the comment commit" in https://github.com/kubernetes/kubernetes/pull/70211#pullrequestreview-171869651 as squashing new changes in an "address comments:" commit, so that we don't have to review the functionality part from the beginning. Please let me know if you want me to squash everything together.

liggitt commented:
lgtm

once @sttts acks, can squash the review commits in and we can tag for merge

sttts commented:
Lgtm. Leaving https://github.com/kubernetes/kubernetes/pull/70211#discussion_r231439381 to @liggitt.

/approve

roycaihw committed:
Add feature gate and API changes for multi-version

schema, subresources and columns

roycaihw committed:
Validation, defaulting and pre-create/update strategy for

CRD per-version schema/subresources/columns

roycaihw committed:
API server serves CRD per-version schema/subresources/columns

roycaihw committed:
Integration test (please review with -w to ignore

whitespace changes)

roycaihw committed:
generated

roycaihw commented:
Commits squashed. Please take a look

liggitt commented:
/lgtm
/approve
